### PR TITLE
add zext and trunc to builder

### PR DESCRIPTION
**Pull request by cqc-melf**
This could be the solution to https://github.com/qir-alliance/pyqir/issues/290
I am not sure if I might have missed anything.

Happy about any feedback.

### CODE REVIEW

**Comment by cqc-melf** [comment]
@microsoft-github-policy-service agree company="Quantinuum"

File: pyqir/src/builder.rs [hunk 1]
```diff
@@ -359,6 +359,36 @@ impl Builder {
...
+
+    /// The ‘zext’ instruction zero extends its operand to the given type.
+    ///
+    /// :param Value val: Value to be converted.
+    /// :param Type ty: Target type.
+    /// :returns: The result.
+    /// :rtype: Value
+    #[pyo3(text_signature = "(self, val, ty)")]
+    fn zext(&self, py: Python, val: &Value, ty: &Type) -> PyResult<PyObject> {
+        let owner = Owner::merge(py, [&self.owner, val.owner(), ty.owner()])?;
```
Comment (idavis) [inline pyqir/src/builder.rs:371]: Type doesn't have an owner.

File: pyqir/src/builder.rs [hunk 2]
```diff
@@ -359,6 +359,36 @@ impl Builder {
...
     }
+
+    /// The ‘zext’ instruction zero extends its operand to the given type.
+    ///
+    /// :param Value val: Value to be converted.
+    /// :param Type ty: Target type.
+    /// :returns: The result.
+    /// :rtype: Value
+    #[pyo3(text_signature = "(self, val, ty)")]
+    fn zext(&self, py: Python, val: &Value, ty: &Type) -> PyResult<PyObject> {
```
Comment (idavis) [inline pyqir/src/builder.rs:370]: Corresponding calls need to be added to `pyqir/pyqir/_native.py` showing the python signature and docs. This is needed as python tooling can't pull the native docs when using the language service, so we have to duplicate the docs.

File: pyqir/src/builder.rs [hunk 3]
```diff
@@ -359,6 +359,36 @@ impl Builder {
...
+
+    /// The ‘trunc’ instruction truncates its operand to the given type.
+    ///
+    /// :param Value val: Value to be converted.
+    /// :param Type ty: Target type.
+    /// :returns: The result.
+    /// :rtype: Value
+    #[pyo3(text_signature = "(self, val, ty)")]
+    fn trunc(&self, py: Python, val: &Value, ty: &Type) -> PyResult<PyObject> {
+        let owner = Owner::merge(py, [&self.owner, val.owner(), ty.owner()])?;
```
Comment (idavis) [inline pyqir/src/builder.rs:386]: As noted above, ty doesn't have an owner.

**Review by idavis — CHANGES_REQUESTED** [review]

File: pyqir/src/builder.rs (hunk 3 above)
Comment (cqc-melf) [reply pyqir/src/builder.rs]: Done in [7e221c8](https://github.com/qir-alliance/pyqir/pull/291/commits/7e221c8bfc52745e18ff5d16b725e2c9ca23aaa9)

File: pyqir/src/builder.rs (hunk 1 above)
Comment (cqc-melf) [reply pyqir/src/builder.rs]: Done in [7e221c8](https://github.com/qir-alliance/pyqir/pull/291/commits/7e221c8bfc52745e18ff5d16b725e2c9ca23aaa9)

File: pyqir/src/builder.rs (hunk 2 above)
Comment (cqc-melf) [reply pyqir/src/builder.rs:370]: Thank you! Done in [b57284a](https://github.com/qir-alliance/pyqir/pull/291/commits/b57284a59fa3e4e96a2f10a5a939e1c1b6a6902c)

**Comment by cqc-melf** [comment]
Thank you for the comments, I have update everything and I hope the checks work now

**Review by idavis — APPROVED** [review]

**Pull request merged**